File: solvers/puzzle_solver.py
import numpy as np
from time import perf_counter as pf

grid = np.zeros((50, 9, 9), dtype=np.uint8)
f = open("puzzles.txt")
m = i = j = 0
for i, row in enumerate(f):
	if row[0]!='G':
		for j, c in enumerate(row[:-1]):
			grid[i//10, i%10-1, j] = int(c)
f.close()
# exit()

def is_available(r, c, val, grid):
	if np.any(grid[r]==val):
		return False
	if np.any(grid[:, c]==val):
		return False
	r = r - r%3
	c = c - c%3
	if np.any(grid[r:r+3, c:c+3]==val):
		return False

	return True

def solve_cell(r, c, grid):
	global solved
	if (r==9):
		return
	r_new = r+1 if c==8 else r
	c_new = (c+1)%9
	if (grid[r, c]):
		solve_cell(r_new, c_new, grid)
	else:
		for val in range(9, 0, -1):
			if is_available(r, c, val, grid):
				grid[r, c] = val
				solve_cell(r_new, c_new, grid)
				if solved:
					return
		if grid[rlast, clast]:
			solved = True
		else:
			grid[r, c] = 0

def check_grid(grid):
	if np.any(np.sum(grid, axis=0)!=45) or np.any(np.sum(grid, axis=1)!=45):
		return False

	# 3x3 grid check left
	return True

solved = False
rlast = clast = 0
found = False
count = 0
t = pf()
for m in range(50):
	t1 = pf()
	solved = False
	rlast = clast = 0
	found = False
	for i in range(8, -1, -1):
		for j in range(8, -1, -1):
			if not grid[m, i, j]:
				rlast = i
				clast = j
				found = True
				break
		if found:
			break
	solve_cell(0, 0, grid[m])
	if check_grid(grid[m]):
		count += 1
		print(f"{m+1}. Solved in {round(pf()-t1, 4)} seconds. ({count} / 50)")
	else:
		print("Not solved. F")
t = pf()-t
print(f"\nAll solved in {t} seconds.")

File: solvers/puzzle_solver.cpp
#include<iostream>
#include<chrono>
#include<fstream>

using namespace std;
#define vvi vector<vector<int>>


int rlast, clast;
long iters=0;
bool solved=false;
int sdku[9][9];

bool is_available(int r, int c, int val) {
    for (int x=0; x<9; ++x) {
        if (sdku[x][c]==val)
            return false;
        if (sdku[r][x]==val)
            return false;
    }
    int gr=r-r%3, gc=c-c%3;
    for (int i=gr; i<gr+3; ++i)
        for (int j=gc; j<gc+3; ++j)
            if (i==r || j==c)
                continue;
            else if (sdku[i][j]==val)
                return false;
    return true;
}
void solve_cell(int r, int c)   {
    if (r==9)
        return;
    ++iters;
    int r_new=(c==8)?r+1:r, c_new=(c+1)%9;
    if (sdku[r][c])
        solve_cell(r_new, c_new);
    else    {
        for (int val=9; val>0; --val)  {
            if (is_available(r, c, val))    {
                sdku[r][c] = val;
                solve_cell(r_new, c_new);
                if (solved)
                    return;
            }
        }
        if (sdku[rlast][clast])
            solved = true;
        else
            sdku[r][c] = 0;
    }
}
bool check_sdku()   {
    for (int j=0; j<9; ++j) {
        int sum1=0, sum2=0;
        for (int i=0; i<9; ++i) {
            sum1 += sdku[i][j];
            sum2 += sdku[j][i];
        }
        if (sum1!=45 || sum2!=45)
            return false;
    }
    // 3x3 grid check left
    return true;
}
void pretty_print() {   // for priting nicely
    for (int i=0; i<9; ++i) {
        for (int j=0; j<9; ++j) {
            cout << sdku[i][j] << " ";
            if (j%3 == 2 && j != 8)
                cout << "| ";
        }
        cout << endl;
        if (i%3 == 2 && i != 8)
                cout << "---------------------\n";
    }
    cout << "\n\n";
}

bool solve()    {
    for (int i=8; i>=0; --i)    {
        for (int j=8; j>=0; --j)
            if (!sdku[i][j])    {
                rlast=i;
                clast=j;
                goto k;
            }
    }
    k:
    solve_cell(0, 0);
    return check_sdku();
}


int main()  {
    ifstream puzzles("puzzles.txt");
    string row;
    int solve_count=0, nump=50;     // nump is 50 at max
    auto t1 = chrono::high_resolution_clock::now();
    
    for (int p=0; p<nump; ++p)    {
        auto t3 = chrono::high_resolution_clock::now();
        // loop start
        solved = false;

        // reading
        for (int i=-1; i<9; ++i)    {
            getline(puzzles, row);
            if (row[0]=='G')
                continue;
            for (int j=0; j<9; ++j)
                sdku[i][j] = row[j]-'0';
        }

        // solving
        bool is_solved = solve();
        auto t4 = chrono::high_resolution_clock::now();
        auto tt = chrono::duration_cast<chrono::microseconds>(t4-t3);
        if (is_solved)    {
            ++solve_count;
            printf("Solved! (%d / %d)\tTime taken: %d microseconds\n", solve_count, nump, tt.count());
        }
        else    {
            cout << "Not solved :(\n";
        }

        // displaying
        // pretty_print();

        // loop end
    }
    auto t2 = chrono::high_resolution_clock::now();
    auto t = chrono::duration_cast<chrono::microseconds>(t2-t1);
    printf("\nSolved %i/%i puzzles in %d microseconds", solve_count, nump, t.count());
    cout << " and " << iters << " total iterations.\n";
    getchar();
}
